tests/recovery/blobmanipulator/blobmanipulator.go

```diff
@@ -0,0 +1,269 @@
+func NewBlobManipulator(baseDirPath string) (*BlobManipulator, error) {
+	runner, err := kopiarunner.NewKopiaSnapshotter(baseDirPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &BlobManipulator{
+		KopiaCommandRunner: runner,
+	}, nil
+}
+
+// ConnectOrCreateRepo connects to an existing repository if possible or creates a new one.
+func (bm *BlobManipulator) ConnectOrCreateRepo(dataRepoPath string) error {
+	return bm.DirCreater.ConnectOrCreateRepo(dataRepoPath)
+}
+
+// DeleteBlob deletes the provided blob or a random blob, in kopia repo.
+func (bm *BlobManipulator) DeleteBlob(blobID string) error {
+	if blobID == "" {
+		randomBlobID, err := bm.getBlobIDRand()
+		blobID = randomBlobID
+
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Printf("Deleting BLOB %s", blobID)
+
+	_, _, err := bm.KopiaCommandRunner.Run("blob", "delete", blobID, "--advanced-commands=enabled")
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (bm *BlobManipulator) getBlobIDRand() (string, error) {
+	var b []blob.Metadata
+
+	// assumption: the repo under test is in filesystem
+	err := bm.KopiaCommandRunner.ConnectRepo("filesystem", "--path="+bm.DataRepoPath)
+	if err != nil {
+		return "", err
+	}
+
+	blobIDList, _, _ := bm.KopiaCommandRunner.Run("blob", "list", "--json")
+	if blobIDList == "" {
+		return "", robustness.ErrNoOp
+	}
+
+	err = json.Unmarshal([]byte(blobIDList), &b)
+	if err != nil {
+		return "", err
+	}
+
+	blobToBeDeleted := ""
+	// Select the first pack blob in the list
+	for _, s := range b {
+		temp := string(s.BlobID)
+		if strings.HasPrefix(temp, "p") {
+			blobToBeDeleted = string(s.BlobID)
+			break
+		}
+	}
+
+	return blobToBeDeleted, nil
+}
+
+func (bm *BlobManipulator) getFileWriter() bool {
+	fw, err := fiofilewriter.New()
+	if err != nil {
+		if errors.Is(err, fio.ErrEnvNotSet) {
+			log.Println("Skipping recovery tests because FIO environment is not set")
+		} else {
+			log.Println("Error creating fio FileWriter:", err)
+		}
+
+		return false
+	}
+
+	bm.fileWriter = fw
+
+	return true
+}
+
+// RestoreGivenOrRandomSnapshot restores a given or a random snapshot from kopia repository into the provided target directory.
+func (bm *BlobManipulator) RestoreGivenOrRandomSnapshot(snapID, restoreDir string) (string, error) {
+	err := bm.KopiaCommandRunner.ConnectRepo("filesystem", "--path="+bm.DataRepoPath)
+	if err != nil {
+		return "", err
+	}
+
+	if snapID == "" {
+		// list available snaphsots
+		stdout, _, snapshotListErr := bm.KopiaCommandRunner.Run("snapshot", "list", "--json")
+		if snapshotListErr != nil {
+			return stdout, snapshotListErr
+		}
+
+		var s []snapshot.Manifest
+
+		err = json.Unmarshal([]byte(stdout), &s)
+		if err != nil {
+			return "", err
+		}
+
+		// Select a random snapshot to restore
+		snapID = string(s[rand.Intn(len(s))].ID)
+	}
+
+	_, msg, err := bm.KopiaCommandRunner.Run("snapshot", "restore", snapID, restoreDir)
+	if err != nil {
+		return msg, err
```

Comment:
Would it be better to return the deleted snap ID from this function, rather of the full `stdout`? that way, the concern of parsing the output remains contained. Either way, can we please document the output name/usage?

Reply:
I had the same thought. However, the `snapshot fix` command does not accept the `snapID` as an argument. It needs the `object ID` mentioned in `stdout`.